### DISCUSSION TITLE
Pressing Ctrl+Alt+<number> / Ctrl+Shift+<number> shouldn't change the ruleset

### DISCUSSION THREAD

**Discussion by Susko3** [ppy/osu#15002] [Q&A]
The intended way to change rulesets is to use `Ctrl`+`<number keys 1-9>`.

But the code only checks if `Ctrl` is pressed, and doesn't check for other modifier keys.
https://github.com/ppy/osu/blob/332cb74cad9f3adf3934432c344ce275be3405ee/osu.Game/Overlays/Toolbar/ToolbarRulesetSelector.cs#L112

The fix is to add `!e.ShiftPressed && !e.AltPressed && !e.SuperPressed`.

`Ctrl`+`Alt`+`<number>` taking priority over text input is a problem in https://github.com/ppy/osu/issues/6338.

---

Making hotkeys with `<modifier key>`+`<number>` seems to be a common thing. But when it's [used](https://github.com/ppy/osu/blob/a5b07ce4feaf88cf2f2f5eee6e43b166a8207737/osu.Game/Overlays/ChatOverlay.cs#L354), checking if other modifier keys are depressed isn't done. Resulting in unwanted behavior.

I think that `OnlyControlPressed`, `OnlyAltPressed`, etc. would be useful to avoid such problems in the future.

---

But what if `Ctrl`+`Shift`+`<number key>` is wanted for a hotkey in the future?
```cs
// obviously doesn't work
e.OnlyControlPressed && e.OnlyShiftPressed

// seems like a bit cumbersome way to write it
e.ControlPressed && e.ShiftPressed && !e.AltPressed && !e.SuperPressed
```

 Something like `ModifiersPressed(Ctrl, Shift)` (or similar) might be a better solution.
(Adding this is currently unnecessary, and is just a nice-to-have.)

---

After we change our usages not to use `Ctrl`+`Alt`, we can make a simple change to allow `AltGr` to work in text boxes framework-side.

```diff
diff --git a/osu.Framework/Graphics/UserInterface/TextBox.cs b/osu.Framework/Graphics/UserInterface/TextBox.cs
index e0fd9f7ab..24919c5ff 100644
--- a/osu.Framework/Graphics/UserInterface/TextBox.cs
+++ b/osu.Framework/Graphics/UserInterface/TextBox.cs
@@ -774,7 +774,9 @@ protected override bool OnKeyDown(KeyDownEvent e)
         {
             if (textInput?.ImeActive == true || ReadOnly) return true;

-            if (e.ControlPressed || e.SuperPressed || e.AltPressed)
+            if ((e.ControlPressed || e.SuperPressed || e.AltPressed)
+                // importantly, don't block Ctrl+Alt (AltGr) as it's used for accented characters on some keyboards.
+                && !(e.ControlPressed && e.AltPressed))
                 return false;
```

**Comment by Susko3** [comment]
The last change can be further refined to _Left_ Control and _Right_ Alt, as AltGr is recognized as those two keys in SDL (at least on Windows).

I know that it's scary to change this (https://github.com/ppy/osu/issues/3125#issuecomment-912379712) but I think a band-aid fix like this should be applied.
Nobody would use LCtrl+RAlt to trigger a Ctrl+Alt hotkey, so it should be a safe change.

**Comment by bdach** [comment]
this should be probably achieved using platform actions and `KeyBindingContainer` rather than peppering in modifier checks. the reason for that would be to accomodate macos (i think tab switching w/ numbers uses <kbd>Cmd</kbd> there? not sure)

either that or `ControlPressed` should change meaning on macos but i'm not that well versed to be sure that that is the correct thing to do.

**Reply by peppy** [reply]
adding to `GlobalAction` sounds good. although that will require special handling for new rulesets, so may be a bit of effort.

**Reply by bdach** [reply]
was thinking `PlatformAction` even (framework-side) and something like `Document{0-9}` or `Tab{0-9}` which would handle new rulesets out of the box.

**Reply by peppy** [reply]
Yeah that may work. I guess it is quite a common behaviour these days.

**Reply by Susko3** [reply]
For osu!, we'd need two sets of actions, as we use Ctrl+number for rulesets, and Alt+number for chat tabs. Or should both have Ctrl+number and have chat take priority?

**Reply by bdach** [reply]
I'd personally be fine with chat tabs taking priority over global ruleset selection. The <kbd>Alt</kbd>+number ones are a bit of a weird case anyhow.

**Reply by Susko3** [reply]
Web browsers use `Ctrl+{1-8}` for the first 8 tabs, I presume you want this to map to `Tab{0-7}`.
`Ctrl+9` is usually mapped to last tab. I think last tab might be useful for chat (less so for rulesets).

Web browsers avoid `Ctrl+0` as it's used for resetting zoom to 100%. We can safely use this for last tab.

Does mapping `Ctrl+{1-9}` to `Tab{0-8}` and `Ctrl+0` to `TabLast` seem fine?
Respectively `Command+{1-0}` for macOS.

**Reply by bdach** [reply]
i think it may be saner to just add `{0-9}` and let implementors decide how to interpret the binding? because it is sometimes context-dependent as you described.
